File: app/src/main/java/hr/caellian/notestream/data/playable/PlayableSource.java
package hr.caellian.notestream.data.playable;

import android.util.Log;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.Objects;

/**
 * Created by tinsv on 03/07/2017.
 */

public enum PlayableSource {
    LOCAL("local", PlayableLocal.class),
    YOUTUBE("remote-youtube", PlayableYouTube.class);

    private static final String TAG = PlayableSource.class.getSimpleName();

    String id;
    Class<? extends Playable> constructorClass;

    PlayableSource(String id, Class<? extends Playable> constructorClass){
        this.id = id;
        this.constructorClass = constructorClass;
    }

    public static PlayableSource getByID(String id) {
        for (PlayableSource source : values()) {
            if (Objects.equals(source.id, id.toLowerCase())) return source;
        }
        Log.e(TAG, "getByID: ", new IllegalArgumentException("Unsupported source ID: '" + id.toLowerCase() + "'!"));
        return null;
    }

    public Playable construct(Object... args) throws IllegalAccessException, InvocationTargetException, InstantiationException {
        ArrayList<Class<?>> argTypes = new ArrayList<>(args.length);
        for (Object arg : args) {
            argTypes.add(arg.getClass());
        }

        Constructor<? extends Playable> constructor;
        try {
            constructor = constructorClass.getConstructor(argTypes.toArray(new Class<?>[0]));
        } catch (NoSuchMethodException e) {
            Log.e(TAG, "construct: ", new IllegalArgumentException("Unsupported argument list!"));
            return null;
        }

        return constructor.newInstance(args);
    }

    public String getID() {
        return id;
    }
}

File: app/src/main/java/hr/caellian/notestream/data/playable/PlayableSource.kt
/*
 * Copyright (C) 2018 Tin Svagelj
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
 * persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
 * NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
 * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package hr.caellian.notestream.data.playable

import android.util.Log
import hr.caellian.notestream.NoteStream
import hr.caellian.notestream.R

enum class PlayableSource(val id: String, private val displayName: Int) {
    LOCAL("local", R.string.location_local),
    YOUTUBE("remote-youtube", R.string.location_youtube),
    OTHER("unknown", R.string.location_unknown);

    fun localizedDisplayName(): String {
        return NoteStream.instance?.getString(this.displayName) ?: ""
    }

    companion object {
        fun getByID(id: String): PlayableSource? {
            values().filter { it.id == id.toLowerCase() }
                    .forEach { return it }

            Log.e(PlayableSource::class.java.simpleName, "getByID: ", IllegalArgumentException("Unsupported source id: '" + id.toLowerCase() + "'!"))
            return null
        }
    }
}
